pkg/auth/requests/impersonate.go

```diff
@@ -99,6 +103,18 @@ func (i *ImpersonatingAuth) ImpersonationMiddleware(next http.Handler) http.Hand
 						util.WriteError(rw, http.StatusForbidden, fmt.Errorf("not allowed to impersonate extras"))
 						return
 					}
+
+					if requestTokenID := reqExtras[common.ExtraRequestTokenID]; len(requestTokenID) > 0 {
+						token, err := i.tokenCache.Get(requestTokenID[0])
```

Comment:
can we check all extras in `common.ExtraRequestTokenID` and not only the first one (requestTokenID[0])? There might be multiple extras with the same name. For example:

- Impersonate-Extra-requesttokenid=goodToken
- Impersonate-Extra-requesttokenid=maliciousToken

In this case we only check the first token which belongs to the user. I think it is worth checking all `Impersonate-Extra-requesttokenid` even if you only take the first value in the imperative API. This will be used for more imperative APIs in the future, and they might get another value that's not the first one. 

For normal requests it won't do any additional check as it will only have one `Impersonate-Extra-requesttokenid`

Reply:
This extra `ExtraRequestTokenID` is meant to be singular by design. Having two or more tokens that are bound to different auth providers and or principals do us no good, not at the moment anyway. That being said I see a potential for misuse in an imperative api store or somewhere else for that matter. What we can do is to reject request outright if it contains more than one value.

Reply:
> This extra `ExtraRequestTokenID` is meant to be singular by design. Having two or more tokens that are bound to different auth providers and or principals do us no good, not at the moment anyway. That being said I see a potential for misuse in an imperative api store or somewhere else for that matter. What we can do is to reject request outright if it contains more than one value.

That makes total sense! More than one value is a request that's impersonating. More than one value is an error or someone trying to escalate permissions since only one value should be used.